### PR TITLE
feat(table): add prop to allow search to always be open

### LINKED ISSUE #3186: [Table] Create prop for an always "open" search field.
### What package is this for?

- [x] React
- [ ] Angular

### Summary

https://next.carbon-addons-iot-react.com/?path=/story/1-watson-iot-table-statefultable--stateful-example-with-row-nesting-and-fixed-columns

Can we add a prop to have the search field open at all times on a data table to align with [this guidance](https://pages.github.ibm.com/ai-applications/design/components/datatable/general#search). This saves users a step looking for and "opening" the search field via the icon button on default tables, while also freeing up an icon button on the right for another use. 

image

### PR DESCRIPTION
**Pull request by kevinsperrine**
Closes #3186 

**Summary**

- adds `isExpanded` prop to Table search props to allow the search field to always be open

**Change List (commits, features, bugs, etc)**

- updated prop types
- added isExpanded prop
- added isExpanded knob to table playground story
- added tests to confirm
- updated snapshots

**Acceptance Test (how to verify the PR)**

- go to the playground story
- turn on the `isExpanded` knob on the Search tab
- confirm the search input stays open regardless of focus or content
- turn off the `isExpanded` knob. Confirm search input opens on focus, stays open with content, and closes when there is no content and no focus

**Regression Test (how to make sure this PR doesn't break old functionality)**

- confirm search continues to function as expected

**Things to look for during review**

- [ ] Make sure all references to `iot` or `bx` class prefix is using the prefix variable
- [ ] (React) All major areas have a `data-testid` attribute. New test ids should have test written to ensure they are not changed or removed.
- [ ] UI should be checked in RTL mode to ensure the proper handling of layout and text.
- [ ] All strings should be translatable.
- [ ] The code should pass a11y scans (The storybook a11y knob should show no violations). New components should have a11y test written.
- [ ] Unit test should be written and should have a coverage of 90% or higher in all areas.
- [ ] All components should be passing visual regression test. For new styles or components either a visual regression test should be written for all permutations or the base image updated.
- [ ] Changes or new components should either write new or update existing documentation.
- [ ] PR should link and close out an existing issue

### CODE REVIEW

**Comment by bjornalm** [comment]
Not sure if this PR was ready for review of not, but the search input does not stay open if I type something and then execute a search or just make the input loose fokus

1. Got to https://deploy-preview-3274--carbon-addons-iot-react.netlify.app/?path=/story/1-watson-iot-table--playground
2. Enable knob hasSearch and isExpanded

image

**Comment by stuckless** [comment]
Quick question about storybook implementation.  When I toggle the isExpanded or the hasSearch it appears to do a full table refresh.  Does it mount an unmount the table to get the table to react to the prop change, or is there something else going on.  ie, I would expect that if I changed the isExanded prop it would be more fluid without a full refresh.

**Comment by kevinsperrine** [comment]
> Quick question about storybook implementation. When I toggle the isExpanded or the hasSearch it appears to do a full table refresh. Does it mount an unmount the table to get the table to react to the prop change, or is there something else going on. ie, I would expect that if I changed the isExanded prop it would be more fluid without a full refresh.

That's a good question. I _believe_ it's related to our work to cleanup our stories and make them both easier to develop and easier to follow for consumers. @bjornalm Is turning knobs on and off causing the `getTableKnobs` function to return and essentially re-rendering the Table each time? I hadn't really thought about it until now.

**Comment by kevinsperrine** [comment]
> Not sure if this PR was ready for review of not, but the search input does not stay open if I type something and then execute a search or just make the input loose fokus
> 
> 1. Got to https://deploy-preview-3274--carbon-addons-iot-react.netlify.app/?path=/story/1-watson-iot-table--playground
> 2. Enable knob hasSearch and isExpanded

@bjornalm Thanks. I had been testing again the Table locally, not StatefulTable, so I had to adjust something in the reducer. Fixed now.

**Comment by bjornalm** [comment]
> > Quick question about storybook implementation. When I toggle the isExpanded or the hasSearch it appears to do a full table refresh. Does it mount an unmount the table to get the table to react to the prop change, or is there something else going on. ie, I would expect that if I changed the isExanded prop it would be more fluid without a full refresh.
> 
> That's a good question. I _believe_ it's related to our work to cleanup our stories and make them both easier to develop and easier to follow for consumers. @bjornalm Is turning knobs on and off causing the `getTableKnobs` function to return and essentially re-rendering the Table each time? I hadn't really thought about it until now.

There are only two knobs that explicitly regenerates the Table key in order for it to update, but this seem to happen for every knob change when the knob `Type of Table` is set to StatefulTable. It could be something in the StatefulTable itself, or a consequence of the massive amount of knobs we are using in our new stories. I will investigate.

**Review by bjornalm — APPROVED** [review]
Nice work, if you update the mdx this is good to go

File: packages/react/src/components/Table/TablePropTypes.js [hunk 1]
```diff
@@ -292,6 +292,7 @@ export const TableSearchPropTypes = PropTypes.shape({
   defaultExpanded: PropTypes.bool,
   onChange: PropTypes.func,
   onExpand: PropTypes.func,
+  isExpanded: PropTypes.bool,
```
[the remark this answers is not in the document]
Comment (kevinsperrine) [reply packages/react/src/components/Table/TablePropTypes.js:295]: Thanks! Added to the docs

File: packages/react/src/components/Table/TableToolbar/TableToolbar.jsx [hunk 2]
```diff
@@ -369,6 +370,7 @@ const TableToolbar = ({
               disabled={isDisabled}
               // TODO: remove deprecated 'testID' in v3
               data-testid={`${testID || testId}-search`}
+              expanded={searchIsExpanded === true ? true : undefined}
```
[the remark this answers is not in the document]
Comment (kevinsperrine) [reply packages/react/src/components/Table/TableToolbar/TableToolbar.jsx:373]: added. that's simpler.

**Comment by bjornalm** [comment]
@kevinsperrine @stuckless I think I figured out why some stories are so slow. 
It happens when stories with large complex components like the StatefulTable are placed within a react wrappers, e.g.  

`MyStory.decorators = [createElement]` or  
`MyStory.decorators = [(Story) => ( <DragAndDrop><Story /></DragAndDrop>)]`
 
 Created an issue and working on a fix:
https://github.com/carbon-design-system/carbon-addons-iot-react/issues/3304

**Comment by davidicus** [comment]
This looks good but do we want the black underline on the search field when it's open ?

**Review by davidicus — APPROVED** [review]

**Comment by bjornalm** [comment]
I'm gonna mention this new props in the new mdx for https://github.com/carbon-design-system/carbon-addons-iot-react/pull/3345 already

**Pull request merged**